puree/src/main/java/com/mercari/puree/TagPattern.java

```diff
@@ -0,0 +1,96 @@
+package com.mercari.puree;
+
+import java.util.regex.Pattern;
+
+import javax.annotation.Nullable;
+
+public class TagPattern {
+
+    private static final String SEPARATOR = "\\.";
+    private static final String ALL_WILD_CARD = "**";
+    private static final String WILD_CARD = "*";
+    private static final Pattern VALID_PATTERN_ALPHANUMERIC = Pattern.compile("^[A-Za-z0-9]+");
+
+    private String pattern;
+
+    public TagPattern() {
+        this.pattern = ALL_WILD_CARD;
+    }
+
+    private TagPattern(String pattern) {
+        this.pattern = pattern;
+    }
+
+    /**
+     * Patterns should follow the following pattern:
+     * • No consecutive periods
+     * • Accepts only alphanumeric, '*', '**', and '.'
+     * @param pattern to be matched
+     * @return whether pattern is valid
+     */
+    private static boolean isValidPattern(String pattern) {
+        String[] patternElements = pattern.split(SEPARATOR);
+        for (String patternPath : patternElements) {
+            if (!VALID_PATTERN_ALPHANUMERIC.matcher(patternPath).matches() &&
+                    !patternPath.equals(ALL_WILD_CARD) &&
+                    !patternPath.equals(WILD_CARD)) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    /**
+     * Matches tag based on the pattern specified in TagPattern
+     *
+     * @param tag
+     * @return whether tag matches pattern scheme
+     */
+    boolean match(String tag) {
+        if (tag.equals(pattern)) {
+            return true;
+        }
+        String[] patternElements = pattern.split(SEPARATOR);
+        String[] tagElements = tag.split(SEPARATOR);
+        int patternIndex = 0;
+        int tagIndex = 0;
+        boolean isAllWildCard = false;
+        while (tagIndex < tagElements.length && patternIndex < patternElements.length) {
```

Comment:
Iterate both zip(patternElements, tagElements) items.

Reply:
Can you give me an example? Which Java 8 function is this?

Reply:
This is pseudo code, but your code works so let's keep it like that for now.